### PR TITLE
Cache user input questions if the question state isn't "complete" and remove the "Settings" link.

### PR DESCRIPTION
**Pull request by tofumatt**
## Summary

Addresses issue #2844.

## Relevant technical choices

Rather than implement a custom TTL setting/logic and also safe storage checks, this re-uses our caching module—which already has flexible/configurable backend support and cache TTL logic.

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/googlesitekit/datastore/user/user-input-settings.js [hunk 1]
```diff
@@ -65,30 +75,50 @@ const baseInitialState = {
...
+	 * @since n.e.x.t
 	 *
-	 * @param {Object} values User input settings.
-	 * @return {Object} Redux-style action.
+	 * @return {Object} Cached user input answer values.
 	 */
-	setUserInputSettings( values ) {
-		return {
-			type: SET_USER_INPUT_SETTINGS,
-			payload: { values },
```
Comment (tofumatt) [inline assets/js/googlesitekit/datastore/user/user-input-settings.js:78]: Note: this was _not_ working as-expected in my usage… using it to set all values resulted in bugs. It's possible the API wasn't clear/documented incorrectly, but more-to-the-point: we don't use this anywhere in the code and it wasn't even useful for this PR really, so I've removed it as:

1. we weren't using it
2. it wasn't working as-expected

The second point matters more, as it led me down a bit of a rabbit hole wondering why the code didn't work. Upon switching to `setUserInputSetting` it worked immediately, so I figured it was better to remove unused, unexpected code.

**Review by eugene-manuilov — APPROVED** [review]
Overall, everything looks good. The only thing that I am not sure about is whether we need to rely on the user input state flag. It has been requested neither in AC nor IB, so I leave it to @felixarntz to judge here.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
@tofumatt could you please check the broken js test?

**Comment by tofumatt** [comment]
> The only thing that I am not sure about is whether we need to rely on the user input state flag.

Yeah, it wasn't in the ACs or IB, but I think it's the best way to handle the logic for storing temporary values. It'll still work if we store them all the time, but it's wasteful so I figured I'd only store when the state isn't "complete".

I've fixed the tests and this is good-to-go now, so I've moved it to merge review given the previous approval. 😄

File: assets/js/googlesitekit/datastore/user/user-input-settings.js [hunk 2]
```diff
@@ -27,25 +27,33 @@ import isPlainObject from 'lodash/isPlainObject';
...
 	reducerCallback: fetchStoreReducerCallback,
 } );
 
 const fetchSaveUserInputSettingsStore = createFetchStore( {
 	baseName: 'saveUserInputSettings',
-	controlCallback: ( settings ) => API.set( 'core', 'user', 'user-input-settings', { settings } ),
+	controlCallback: ( settings ) => {
+		return API.set( 'core', 'user', 'user-input-settings', { settings } )
+		;
+	},
```
Comment (felixarntz) [inline assets/js/googlesitekit/datastore/user/user-input-settings.js:56]: These changes seem unrelated/unnecessary?

File: assets/js/googlesitekit/datastore/user/user-input-settings.js [hunk 3]
```diff
@@ -65,30 +75,50 @@ const baseInitialState = {
 
 const baseActions = {
 	/**
-	 * Sets user input settings.
+	 * Gets cached user input settings and save them to the data store.
 	 *
-	 * @since 1.19.0
+	 * @since n.e.x.t
```
Comment (felixarntz) [inline assets/js/googlesitekit/datastore/user/user-input-settings.js:80]: Since this action is documented, it should be marked as private, since it's a low-level action that should only be used via the `getUserInputSettings` resolver. The alternative is to not document it at all.

File: assets/js/googlesitekit/datastore/user/user-input-settings.js [hunk 4]
```diff
@@ -65,30 +75,50 @@ const baseInitialState = {
...
 	 *
-	 * @param {Object} values User input settings.
-	 * @return {Object} Redux-style action.
+	 * @return {Object} Cached user input answer values.
 	 */
-	setUserInputSettings( values ) {
-		return {
-			type: SET_USER_INPUT_SETTINGS,
-			payload: { values },
+	*fetchCachedUserInputSettings() {
```
Comment (felixarntz) [inline assets/js/googlesitekit/datastore/user/user-input-settings.js:84]: This name is a bit off for two reasons:
* The `fetch*` prefix should only be used for the internal fetch actions related to a public API-based selector or action (typically using `createFetchStore`). The action here isn't that though.
* The function sets user input settings based on what is cached, which is not clear from the name (fetch implies it's getting/fetching something, but it also sets something).

I suggest something like `setUserInputSettingsFromCache`.

File: assets/js/googlesitekit/datastore/user/user-input-settings.js [hunk 5]
```diff
@@ -65,30 +75,50 @@ const baseInitialState = {
...
-	setUserInputSetting( settingID, values ) {
+	*setUserInputSetting( settingID, values ) {
+		const registry = yield Data.commonActions.getRegistry();
+		if ( registry.select( STORE_NAME ).getUserInputState() !== 'completed' ) {
+			// Save this setting in the cache.
+			yield {
+				type: SET_CACHED_USER_INPUT_SETTING,
+				payload: {
+					settingID,
+					values,
```
Comment (felixarntz) [inline assets/js/googlesitekit/datastore/user/user-input-settings.js:117]: This is missing the `trim` logic that is needed, see right below here.

File: assets/js/googlesitekit/datastore/user/user-input-settings.js [hunk 6]
```diff
@@ -27,25 +27,33 @@ import isPlainObject from 'lodash/isPlainObject';
...
 import Data from 'googlesitekit-data';
+import { deleteItem, getItem, setItem } from '../../../googlesitekit/api/cache';
 import { STORE_NAME } from './constants';
 import { createFetchStore } from '../../data/create-fetch-store';
 import { actions as errorStoreActions } from '../../data/create-error-store';
-const { commonActions, createRegistrySelector } = Data;
+const { commonActions, createRegistryControl, createRegistrySelector } = Data;
 const { receiveError, clearError } = errorStoreActions;
 
+const CACHE_KEY_NAME = 'fetchGetUserInputSettingsStore.fetchGetUserInputSettingsStore';
```
Comment (felixarntz) [inline assets/js/googlesitekit/datastore/user/user-input-settings.js:37]: Why this overly cryptic/specific cache key name with a duplicated segment? It makes me think that name is somehow required due to how it integrates with `fetchGetUserInputSettingsStore`. We can just give it an arbitrary name like `userInputSettings`.

File: assets/js/googlesitekit/datastore/user/user-input-settings.js [hunk 7]
```diff
@@ -173,9 +218,14 @@ export const baseReducer = ( state, { type, payload } ) => {
 const baseResolvers = {
 	*getUserInputSettings() {
 		const { select } = yield commonActions.getRegistry();
-		if ( ! select( STORE_NAME ).getUserInputSettings() ) {
+
+		if ( ! select( STORE_NAME ).getUserInputSettings() && select( STORE_NAME ).getUserInputState() === 'completed' ) {
```
Comment (felixarntz) [inline assets/js/googlesitekit/datastore/user/user-input-settings.js:222]: This seems wrong. There might be saved user input settings also when the state is not `completed`. This part shouldn't be changed, we need to only make sure that the cached user input settings take precedence over what comes from the server.

**Review by felixarntz — CHANGES_REQUESTED** [review]
Overall looks solid; I left a bit of minor feedback, a few things need iteration though.

File: assets/js/googlesitekit/datastore/user/user-input-settings.js (hunk 2 above)
Comment (tofumatt) [reply assets/js/googlesitekit/datastore/user/user-input-settings.js:56]: Ah, I think I was doing something in there with console logs previously, removed them, and that's why the change is there. I've reverted it.

File: assets/js/googlesitekit/datastore/user/user-input-settings.js (hunk 6 above)
Comment (tofumatt) [reply assets/js/googlesitekit/datastore/user/user-input-settings.js:37]: Fair point—there's no reason for it other than specificity, but `userInputSettings` should be fine 👍🏻

File: assets/js/googlesitekit/datastore/user/user-input-settings.js (hunk 7 above)
Comment (tofumatt) [reply assets/js/googlesitekit/datastore/user/user-input-settings.js:222]: Ah, fair point. I guess since we never store it _after_ the user input status is `complete` that's safe to do. 👍🏻

**Review by felixarntz — APPROVED** [review]
Great!
